src/System.IO.Pipelines.Networking.Tls/OpenSslSecurityContext.cs

```diff
@@ -0,0 +1,141 @@
+        internal Interop.alpn_cb AlpnCallback;
+        private ApplicationProtocols.ProtocolIds _alpnSupportedProtocols;
+
+        public OpenSslSecurityContext(PipelineFactory pipelineFactory, string hostName, bool isServer, string pathToPfxFile, string password)
+            : this(pipelineFactory, hostName, isServer, pathToPfxFile, password, 0)
+        {
+        }
+
+        public OpenSslSecurityContext(PipelineFactory pipelineFactory, string hostName, bool isServer, string pathToPfxFile, string password, ApplicationProtocols.ProtocolIds alpnSupportedProtocols)
+        {
+            if (isServer && string.IsNullOrEmpty(pathToPfxFile))
+            {
+                throw new ArgumentException("We need a certificate to load if you want to run in server mode");
+            }
+
+            InteropCrypto.Init();
+            _hostName = hostName;
+            _pipelineFactory = pipelineFactory;
+            _isServer = isServer;
+            _alpnSupportedProtocols = alpnSupportedProtocols;
+
+            if (!string.IsNullOrWhiteSpace(pathToPfxFile))
+            {
+                InteropBio.BioHandle fileBio = new InteropBio.BioHandle();
+                try
+                {
+                    fileBio = InteropBio.BIO_new_file_read(pathToPfxFile);
+                    //Now we pull out the private key, certificate and Authority if they are all there
+                    _certifcateInformation = new InteropKeys.PK12Certifcate(fileBio, password);
+                }
+                finally
+                {
+                    fileBio.FreeBio();
+                }
+            }
+            SetupContext();
+            SetupAlpn();
+        }
+
+        public bool IsServer => _isServer;
+        internal InteropKeys.PK12Certifcate CertificateInformation => _certifcateInformation;
+        internal IntPtr AplnBuffer => _alpnHandle.AddrOfPinnedObject();
+        internal int AplnBufferLength => _alpnSupportedProtocolsBuffer?.Length ?? 0;
+
+        private unsafe void SetupAlpn()
+        {
+            if (_alpnSupportedProtocols > 0)
+            {
+                //We need to get a buffer for the ALPN negotiation and pin it for sending to the lower API
+                _alpnSupportedProtocolsBuffer = ApplicationProtocols.GetBufferForProtocolId(_alpnSupportedProtocols, false);
+                _alpnHandle = GCHandle.Alloc(_alpnSupportedProtocolsBuffer, GCHandleType.Pinned);
+                Interop.SSL_CTX_set_alpn_protos(SslContext, AplnBuffer, (uint)AplnBufferLength);
+                if (_isServer)
+                {
+                    AlpnCallback = alpn_cb;
+                    Interop.SSL_CTX_set_alpn_select_cb(this);
+                }
+            }
+        }
+
+        private unsafe Interop.AlpnStatus alpn_cb(IntPtr ssl, out byte* selProto, out byte selProtoLen, byte* inProtos, int inProtosLen, IntPtr arg)
+        {
+            byte* matchBuffer = (byte*)_alpnHandle.AddrOfPinnedObject();
```

Comment:
wrap that method signature? It's a callback from OpenSsl they don't accept spans or am I misreading you